README.md
[![Build Status](https://travis-ci.org/indigo-astronomy/indigo.svg?branch=master)](https://travis-ci.org/indigo-astronomy/indigo)
[![Tag Version](https://img.shields.io/github/tag/indigo-astronomy/indigo.svg)](https://github.com/indigo-astronomy/indigo/tags)
[![License](http://img.shields.io/badge/INDIGO-license-red.svg)](https://github.com/indigo-astronomy/indigo/blob/master/LICENSE.md)

# INDIGO is the next generation of INDI, based on layered architecture and software bus.

This is the list of requirements taken into the consideration:

1. No GPL dependency to allow commercial use under application stores licenses.
2. ANSI C for portability and to allow simple wrapping into .Net, Java, GOlang, Objective-C or Swift in future
3. Layered architecture to allow direct linking of the drivers into applications and or to replace wire protocol.
4. Atomic approach to device drivers. E.g. if camera has imaging and guiding chip, driver should expose two independent simple devices instead of one complex. It is much easier and transparent for client.
5. Drivers should support hot-plug at least for USB devices. If device is connected/disconnected while driver is running, its properties should appear/disappear on the bus.
6. FITS, XISF, JPEG and RAW format supported directly by the framework.

## This is already done

#### Framework

1. INDIGO Bus framework
2. XML & JSON protocol adapter for client and server side
3. indigo_server - server with loadable (indigo .so/.dylib and indi executables) drivers
4. indigo_server_standalone - server with built-in drivers
5. INDIGO Server for macOS wrapper (provided just as an example)
6. Integrated HTTP server for BLOB download & server control (= web based INDI control panel)
7. indigo_prop_tool - command line tool to list and set properties

#### Drivers

1. CCD (with wheel, guider and focuser) simulator
2. Mount simulator
3. Atik CCD (Titan, 3xx/4xx/One with built in wheel, VS/Infinity, 11000/4000) driver
4. Atik EFW2 filterwheel driver
5. SX CCD driver
6. SX wheel driver
7. Shoestring FCUSB focuser driver
8. SSAG/QHY5 CCD driver
9. ASI wheel driver
10. IIDC CCD driver
11. ASI CCD driver
12. NexStar mount driver (supports Celestron NexStar and Sky-Watcher mounts)
13. LX200 mount driver (supports Meade mounts and EQMac)
14. FLI filter wheel driver
15. FLI CCD driver
16. FLI focuser driver (testers needed)
17. USB_Focus v3 driver
18. SBIG CCD driver (with guider, guider CCD and external guider head)
19. SBIG Filter wheel driver (part of CCD driver)
20. ASCOM driver for INDIGO Camera
21. ASCOM driver for INDIGO Filter Wheel
22. QHY CCD driver (NOTE: Maybe unstable due to inherited instability in the QHY SDK)
23. ZWO USB-ST4 port
24. Meade DSI Camera driver
25. Takahashi Temma mount driver
26. ICA (ImageCapture API) driver for DSLRs (Mac only)
27. GPS Simulator
28. NMEA 0183 GPS driver
29. Andor CCD driver (32/64-bit Intel Linux only)
30. WeMacro Rail focuser driver (platform independent USB & Mac only bluetooth)
31. EQMac guider driver (Mac only)
32. Apogee CCD driver
33. Moravian Intruments CCD and filter wheel driver
34. HID Joystick AUX driver
35. CGUSBST4 guider driver
36. Brightstar Quantum filter wheel driver (untested)
37. Trutek filter wheel driver (untested)
38. Xagyl filter wheel driver (untested)
39. Optec filter wheel driver (untested)
40. Pegasus DMFC focuser driver
41. RigelSys nSTEP focuser driver
42. RigelSys nFOCUS focuser driver (untested)
43. iOptron mount driver
44. MoonLite focuser driver (untested)
45. MJKZZ rail focuser driver (untested, platform independent USB & Mac only bluetooth)
46. GPhoto2 CCD driver
47. Optec focuser driver (untested)
48. ToupTek CCD driver
49. AltairAstro CCD driver
50. RTS-on-COM aux (shutter) driver (untested)
51. DSUSB aux (shutter) driver (untested)
52. GPUSB guider driver (untested)
53. LakesideAstro focuser (untested)

## This is under development
1. ASCOM driver for INDIGO mount
2. ASCOM driver for INDIGO focuser
3. Apogee filter wheel driver
4. SynScan (EQMod) mount driver

------------------------------------------------------------------------------------------------
## How to build INDIGO

### Prerequisits
#### Ubuntu / Debian / Raspbian

`sudo apt-get install build-essential autoconf autotools-dev libtool cmake libudev-dev libavahi-compat-libdnssd-dev libusb-1.0-0-dev fxload libcurl4-gnutls-dev libgphoto2-dev git curl`

It is advised to remove libraw1394-dev

`sudo apt-get remove libraw1394-dev`

#### Fedora

`dnf install automake autoconf cmake libtool gcc gcc-c++ libusb-devel avahi-compat-libdns_sd-devel libudev-devel libgphoto2-devel git curl`

It is advised to remove libraw1394-devel

`dnf remove libraw1394-devel`

#### macOS

Install XCode and download and build autoconf, automake and libtool (use tools/cltools.sh script).

### Get code and build it

`git clone https://github.com/indigo-astronomy/indigo.git`

`cd indigo`

`make all`

`build/bin/indigo_server -v -s`

and connect from any INDIGO/INDI client or web browser to localhost on port 7624...


ChangeLog
Thu Apr  6 15:09:08 2016 - 0.15 (skipping 0.14 to catch up with NexStarCtl)
	- added full SkyWatcher support
	- distinguish between SkyWatcher and Celestron
	- new command tc_get_orientation();
	- documentation fixes
	- several new calls
Sat Sep  5 03:38:23 2015 - 0.13
	- Fix typo
Sat Mar 29 15:45:07 2014 - 0.11
	- added PEC record/playback, download/upload
Wed Mar 12 10:46:06 2014 - 0.10
	- remove some debug messages
	- CGEM mount is reconized
Sun Mar  9 21:10:07 2014 - 0.09
	- fix data len in tc_set_backlash().
Sun Mar  9 16:29:27 2014 - 0.08
	- test fix and document tc_get_autoguide_rate(), tc_set_autoguide_rate(),
	  tc_get_backlash(), tc_set_backlash().
Thu Mar  6 18:22:21 2014 - Untested
	- add named function return codes rather than numbers
	- implemented, but not tested tc_get_autoguide_rate(),
	  tc_set_autoguide_rate(), tc_get_backlash(), tc_set_backlash().
Wed Dec  4 23:37:29 2013 - 0.07
	- change version number to more reasonable notation
	- fix tc_get_time() one hour offset if DST/ST is incorrect
Wed Dec  4 01:22:46 2013
	- read_telescope(): unsigned char -> char
	- fixed some warnings on MacOSX
	- tc_set_time() fix since RTC time expects UT
Tue Dec  3 11:33:19 2013
	- tc_set_time() sets date/time to RTC too, if available
	- updated libnexstar.man
Sun Nov  3 12:58:20 2013 - 0.1-a5
	- fix ID of Advanced VX
Fri Oct 25 14:56:41 2013 - 0.1-a4
	- added get_model_name()
Sat Oct 12 18:24:11 2013 - 0.1-a3
	- made it C++ linkable
Sun Sep 29 16:44:53 2013 - 0.1-a1
	- implemented all the major commands
	- fixed read_telescope()
	- fixed open_telescope() for raw mode
	- fixed declination +90 shown as -270
Wed Sep 25 15:44:42 2013
	-libnexstar: builds as .so and .a
	-implemented several functions
	-not tested yet


dc1394/internal.c
/*
 * 1394-Based Digital Camera Control Library
 *
 * Internal functions
 *
 * Written by Damien Douxchamps <ddouxchamps@users.sf.net>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "internal.h"
#include "utils.h"
#include "log.h"
#include "register.h"

/*
  These arrays define how many image quadlets there
  are in a packet given a mode and a frame rate
  This is defined in the 1394 digital camera spec
*/
const int quadlets_per_packet_format_0[56] = {
    -1,  -1,  15,  30,  60,  120,  240,  480,
    10,  20,  40,  80, 160,  320,  640, 1280,
    30,  60, 120, 240, 480,  960, 1920, 3840,
    40,  80, 160, 320, 640, 1280, 2560, 5120,
    60, 120, 240, 480, 960, 1920, 3840, 7680,
    20,  40,  80, 160, 320,  640, 1280, 2560,
    40,  80, 160, 320, 640, 1280, 2560, 5120
};

const int quadlets_per_packet_format_1[64] =  {
    -1, 125, 250,  500, 1000, 2000, 4000, 8000,
    -1,  -1, 375,  750, 1500, 3000, 6000,   -1,
    -1,  -1, 125,  250,  500, 1000, 2000, 4000,
    96, 192, 384,  768, 1536, 3072, 6144,   -1,
    144, 288, 576, 1152, 2304, 4608,   -1,   -1,
    48,  96, 192,  384,  768, 1536, 3073, 6144,
    -1, 125, 250,  500, 1000, 2000, 4000, 8000,
    96, 192, 384,  768, 1536, 3072, 6144,   -1
};

const int quadlets_per_packet_format_2[64] =  {
    160, 320,  640, 1280, 2560, 5120,   -1, -1,
    240, 480,  960, 1920, 3840, 7680,   -1, -1,
    80, 160,  320,  640, 1280, 2560, 5120, -1,
    250, 500, 1000, 2000, 4000, 8000,   -1, -1,
    375, 750, 1500, 3000, 6000,   -1,   -1, -1,
    125, 250,  500, 1000, 2000, 4000, 8000, -1,
    160, 320,  640, 1280, 2560, 5120,   -1, -1,
    250, 500, 1000, 2000, 4000, 8000,   -1, -1
};

/********************************************************
 get_quadlets_per_packet

 This routine reports the number of useful image quadlets
 per packet
*********************************************************/
dc1394error_t
get_quadlets_per_packet(dc1394video_mode_t mode, dc1394framerate_t frame_rate, uint32_t *qpp) // ERROR handling to be updated
{
    uint32_t mode_index;
    uint32_t frame_rate_index= frame_rate - DC1394_FRAMERATE_MIN;
    uint32_t format;
    dc1394error_t err;

    err=get_format_from_mode(mode, &format);
    DC1394_ERR_RTN(err,"Invalid mode ID");

    switch(format) {
    case DC1394_FORMAT0:
        mode_index= mode - DC1394_VIDEO_MODE_FORMAT0_MIN;

        if ( ((mode >= DC1394_VIDEO_MODE_FORMAT0_MIN) && (mode <= DC1394_VIDEO_MODE_FORMAT0_MAX)) &&
             ((frame_rate >= DC1394_FRAMERATE_MIN) && (frame_rate <= DC1394_FRAMERATE_MAX)) ) {
            *qpp=quadlets_per_packet_format_0[DC1394_FRAMERATE_NUM*mode_index+frame_rate_index];
        }
        else {
            err=DC1394_INVALID_VIDEO_MODE;
            DC1394_ERR_RTN(err,"Invalid framerate or mode");
        }
        return DC1394_SUCCESS;
    case DC1394_FORMAT1:
        mode_index= mode - DC1394_VIDEO_MODE_FORMAT1_MIN;

        if ( ((mode >= DC1394_VIDEO_MODE_FORMAT1_MIN) && (mode <= DC1394_VIDEO_MODE_FORMAT1_MAX)) &&
             ((frame_rate >= DC1394_FRAMERATE_MIN) && (frame_rate <= DC1394_FRAMERATE_MAX)) ) {
            *qpp=quadlets_per_packet_format_1[DC1394_FRAMERATE_NUM*mode_index+frame_rate_index];
        }
        else {
            err=DC1394_INVALID_VIDEO_MODE;
            DC1394_ERR_RTN(err,"Invalid framerate or mode");
        }
        return DC1394_SUCCESS;
    case DC1394_FORMAT2:
        mode_index= mode - DC1394_VIDEO_MODE_FORMAT2_MIN;

        if ( ((mode >= DC1394_VIDEO_MODE_FORMAT2_MIN) && (mode <= DC1394_VIDEO_MODE_FORMAT2_MAX)) &&
             ((frame_rate >= DC1394_FRAMERATE_MIN) && (frame_rate <= DC1394_FRAMERATE_MAX)) ) {
            *qpp=quadlets_per_packet_format_2[DC1394_FRAMERATE_NUM*mode_index+frame_rate_index];
        }
        else {
            err=DC1394_INVALID_VIDEO_MODE;
            DC1394_ERR_RTN(err,"Invalid framerate or mode");
        }
        return DC1394_SUCCESS;
    case DC1394_FORMAT6:
    case DC1394_FORMAT7:
        err=DC1394_INVALID_VIDEO_FORMAT;
        DC1394_ERR_RTN(err,"Format 6 and 7 don't have qpp");
        break;
    }

    return DC1394_FAILURE;
}

/**********************************************************
 get_quadlets_from_format

 This routine reports the number of quadlets that make up a
 frame given the format and mode
***********************************************************/
dc1394error_t
get_quadlets_from_format(dc1394camera_t *camera, dc1394video_mode_t video_mode, uint32_t *quads)
{
    uint32_t w, h, color_coding;
    uint32_t bpp;
    dc1394error_t err;

    err=dc1394_get_image_size_from_video_mode(camera, video_mode, &w, &h);
    DC1394_ERR_RTN(err, "Invalid mode ID");

    err=dc1394_get_color_coding_from_video_mode(camera, video_mode, &color_coding);
    DC1394_ERR_RTN(err, "Invalid mode ID");

    err=dc1394_get_color_coding_bit_size(color_coding, &bpp);
    DC1394_ERR_RTN(err, "Invalid color mode ID");

    *quads=(w*h*bpp)/32;

    return err;
}

dc1394bool_t
is_feature_bit_set(uint32_t value, dc1394feature_t feature)
{

    if (feature >= DC1394_FEATURE_ZOOM) {
        if (feature >= DC1394_FEATURE_CAPTURE_SIZE) {
            feature+= 12;
        }
        feature-= DC1394_FEATURE_ZOOM;
    }
    else {
        feature-= DC1394_FEATURE_MIN;
    }

    value&=(0x80000000UL >> feature);

    if (value>0)
        return DC1394_TRUE;
    else
        return DC1394_FALSE;
}

dc1394error_t
get_format_from_mode(dc1394video_mode_t mode, uint32_t *format)
{
    dc1394error_t err=DC1394_SUCCESS;

    if ((mode>=DC1394_VIDEO_MODE_FORMAT0_MIN)&&(mode<=DC1394_VIDEO_MODE_FORMAT0_MAX)) {
        *format=DC1394_FORMAT0;
    }
    else if ((mode>=DC1394_VIDEO_MODE_FORMAT1_MIN)&&(mode<=DC1394_VIDEO_MODE_FORMAT1_MAX)) {
        *format=DC1394_FORMAT1;
    }
    else if ((mode>=DC1394_VIDEO_MODE_FORMAT2_MIN)&&(mode<=DC1394_VIDEO_MODE_FORMAT2_MAX)) {
        *format=DC1394_FORMAT2;
    }
    else if ((mode>=DC1394_VIDEO_MODE_FORMAT6_MIN)&&(mode<=DC1394_VIDEO_MODE_FORMAT6_MAX)) {
        *format=DC1394_FORMAT6;
    }
    else if ((mode>=DC1394_VIDEO_MODE_FORMAT7_MIN)&&(mode<=DC1394_VIDEO_MODE_FORMAT7_MAX)) {
        *format=DC1394_FORMAT7;
    }
    else {
        err=DC1394_INVALID_VIDEO_MODE;
        DC1394_ERR_RTN(err, "The supplied mode does not correspond to any format");
    }

    return err;
}


dc1394error_t
capture_basic_setup (dc1394camera_t * camera, dc1394video_frame_t * frame)
{
    dc1394error_t err;
    uint32_t bpp;
    dc1394video_mode_t video_mode;
    dc1394framerate_t framerate;

    frame->camera = camera;

    err=dc1394_video_get_mode(camera,&video_mode);
    DC1394_ERR_RTN(err, "Unable to get current video mode");
    frame->video_mode = video_mode;

    err=dc1394_get_image_size_from_video_mode(camera, video_mode, frame->size, frame->size + 1);
    DC1394_ERR_RTN(err,"Could not get width/height from format/mode");

    if (dc1394_is_video_mode_scalable(video_mode)==DC1394_TRUE) {

        err=dc1394_format7_get_packet_size(camera, video_mode,
                                           &frame->packet_size);
        DC1394_ERR_RTN(err, "Unable to get format 7 bytes per packet");

        err=dc1394_format7_get_packets_per_frame(camera, video_mode,
                                                 &frame->packets_per_frame);
        DC1394_ERR_RTN(err, "Unable to get format 7 packets per frame");

        err = dc1394_format7_get_image_position (camera, video_mode,
                                                 frame->position, frame->position + 1);
        DC1394_ERR_RTN(err, "Unable to get format 7 image position");

        dc1394_format7_get_color_filter (camera, video_mode, &frame->color_filter);
    }
    else {
        err=dc1394_video_get_framerate(camera,&framerate);
        DC1394_ERR_RTN(err, "Unable to get current video framerate");

        err=get_quadlets_per_packet(video_mode, framerate, &frame->packet_size);
        DC1394_ERR_RTN(err, "Unable to get quadlets per packet");
        frame->packet_size *= 4;

        err= get_quadlets_from_format(camera, video_mode, &frame->packets_per_frame);
        DC1394_ERR_RTN(err,"Could not get quadlets per frame");
        frame->packets_per_frame /= frame->packet_size/4;

        frame->position[0] = 0;
        frame->position[1] = 0;
        frame->color_filter = 0;
    }

    dc1394_log_debug("Mode %d, %dx%d, packet size %d, "
            "packets per frame %d\n",
            frame->video_mode, frame->size[0], frame->size[1],
            frame->packet_size, frame->packets_per_frame);

    if ((frame->packet_size <=0 )||
        (frame->packets_per_frame <= 0)) {
        return DC1394_FAILURE;
    }

    frame->yuv_byte_order = DC1394_BYTE_ORDER_UYVY;

    frame->total_bytes = frame->packets_per_frame * frame->packet_size;

    err = dc1394_get_color_coding_from_video_mode (camera, video_mode, &frame->color_coding);
    DC1394_ERR_RTN(err, "Unable to get color coding");

    frame->data_depth=0; // to avoid valgrind warnings
    err = dc1394_video_get_data_depth (camera, &frame->data_depth);
    DC1394_ERR_RTN(err, "Unable to get data depth");

    err = dc1394_get_color_coding_bit_size (frame->color_coding, &bpp);
    DC1394_ERR_RTN(err, "Unable to get bytes per pixel");

    frame->stride = (bpp * frame->size[0])/8;
    frame->image_bytes = frame->size[1] * frame->stride;
    frame->padding_bytes = frame->total_bytes - frame->image_bytes;

    frame->little_endian=0;   // not used before 1.32 is out.
    frame->data_in_padding=0; // not used before 1.32 is out.

    return DC1394_SUCCESS;
}

